File: desktop/crates/ccp-protocol/src/control/telemetry.rs
//! Telemetry, ping/pong (§6.4).

use serde::{Deserialize, Serialize};

use crate::control::device::{BatteryState, ThermalState};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Telemetry {
    /// Device monotonic clock, microseconds.
    pub ts_usec: u64,
    pub battery_level: f64,
    pub battery_state: BatteryState,
    pub thermal_state: ThermalState,
    pub sent_bitrate_kbps: u32,
    pub enc_fps: u32,
    pub capture_fps: u32,
    pub queue_depth: u32,
    pub drop_count: u64,
}

// Telemetry contains `f64` so it cannot be `Eq`; only `PartialEq`.

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Ping {
    pub ts_usec: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Pong {
    pub ts_usec: u64,
    pub echo_usec: u64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn telemetry_round_trip() {
        let t = Telemetry {
            ts_usec: 1_000_000,
            battery_level: 0.5,
            battery_state: BatteryState::Charging,
            thermal_state: ThermalState::Fair,
            sent_bitrate_kbps: 25_000,
            enc_fps: 30,
            capture_fps: 30,
            queue_depth: 1,
            drop_count: 2,
        };
        let s = serde_json::to_string(&t).unwrap();
        let back: Telemetry = serde_json::from_str(&s).unwrap();
        assert_eq!(back, t);
    }

    #[test]
    fn ping_pong_round_trip() {
        let p = Ping { ts_usec: 12345 };
        let s = serde_json::to_string(&p).unwrap();
        assert_eq!(serde_json::from_str::<Ping>(&s).unwrap(), p);

        let q = Pong {
            ts_usec: 23456,
            echo_usec: 12345,
        };
        let s = serde_json::to_string(&q).unwrap();
        assert_eq!(serde_json::from_str::<Pong>(&s).unwrap(), q);
    }
}

File: ios/Sources/ClearCamProtocol/Telemetry.swift
import Foundation

public struct Telemetry: Codable, Equatable {
    public var tsUsec: UInt64
    public var batteryLevel: Double
    public var batteryState: BatteryState
    public var thermalState: ThermalState
    public var sentBitrateKbps: UInt32
    public var encFps: UInt32
    public var captureFps: UInt32
    public var queueDepth: UInt32
    public var dropCount: UInt64

    public init(
        tsUsec: UInt64, batteryLevel: Double, batteryState: BatteryState,
        thermalState: ThermalState, sentBitrateKbps: UInt32, encFps: UInt32, captureFps: UInt32,
        queueDepth: UInt32, dropCount: UInt64
    ) {
        self.tsUsec = tsUsec
        self.batteryLevel = batteryLevel
        self.batteryState = batteryState
        self.thermalState = thermalState
        self.sentBitrateKbps = sentBitrateKbps
        self.encFps = encFps
        self.captureFps = captureFps
        self.queueDepth = queueDepth
        self.dropCount = dropCount
    }
}

public struct Ping: Codable, Equatable {
    public var tsUsec: UInt64
    public init(tsUsec: UInt64) { self.tsUsec = tsUsec }
}

public struct Pong: Codable, Equatable {
    public var tsUsec: UInt64
    public var echoUsec: UInt64
    public init(tsUsec: UInt64, echoUsec: UInt64) {
        self.tsUsec = tsUsec
        self.echoUsec = echoUsec
    }
}
